### PR TITLE
Operator restructure

### PR DESCRIPTION
**Pull request by mrogers950**
Use the cluster-osin-operator repo layout and boilerplate (with minor modifications).
@openshift/sig-auth

### CODE REVIEW

File: manifests/0000_51_machine-approver-02-clusterrole.yaml [hunk 1]
```diff
@@ -18,3 +18,25 @@ rules:
   - certificatesigningrequests/approval
   verbs:
   - update
+- apiGroups:
```
Comment (enj) [inline manifests/0000_51_machine-approver-02-clusterrole.yaml]: These changes do not look correct to me.

Even if we need these permissions, we probably do not need them at the cluster scope.

**Comment by mrogers950** [comment]
@enj fixed the roles, and disabled authn/authz delegation in the default config. I also reverted my changes to the boilerplate and used controller.New like you suggested..

File: manifests/0000_51_machine-approver-04-role.yaml [hunk 2]
```diff
@@ -0,0 +1,34 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: machine-approver
+  namespace: openshift-cluster-machine-approver
+rules:
+- apiGroups:
+  - ""
+  resources:
+  - pods
```
Comment (enj) [inline manifests/0000_51_machine-approver-04-role.yaml]: Are all these required because of the controller command's default behavior?

**Comment by enj** [comment]
Going to tag because I have a feeling that changing https://github.com/openshift/cluster-machine-approver/pull/8#discussion_r243133661 will require messing with the controller command bits.

/lgtm
/retest

File: cmd/machine-approver/main.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	if err := command.Execute(); err != nil {
+		fmt.Fprintf(os.Stderr, "%v\n", err)
+		os.Exit(1)
+	}
+}
+
+func NewMachineApproverOperatorCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "machine-approver",
+		Short: "OpenShift osin OAuth server operator",
```
Comment (ericavonb) [inline cmd/machine-approver/main.go]: reminder to update.

**Comment by smarterclayton** [comment]
Since this merged I've seen two failures where the machine-approver-controller is crash looping:

https://openshift-gce-devel.appspot.com/build/origin-ci-test/logs/release-openshift-origin-installer-e2e-aws-4.0/2708
https://openshift-gce-devel.appspot.com/build/origin-ci-test/logs/release-openshift-origin-installer-e2e-aws-4.0/2703

which I had never seen prior to to this.

**Comment by smarterclayton** [comment]
I would have liked to see a lot more green runs before this merged, given that it was a clean controller.  It looks like you got exactly 1 pass after failing 29 times in a row - that alone is an eyebrow raiser.

In general, whenever we rewrite / refactor a controller, please ensure you see multiple green e2e-aws runs before merging.  If I see continued flakes in the next few days, and it looks like this controller is at fault, I will be reverting this (unless you can point to a different cause).

File: cmd/machine-approver/main.go (hunk 3 above)
Comment (smarterclayton) [reply cmd/machine-approver/main.go]: Looks like you didn't fix this.

File: hack/build-rpms.sh [hunk 4]
```diff
@@ -0,0 +1,131 @@
+#!/bin/bash
```
Comment (smarterclayton) [inline hack/build-rpms.sh]: Why do you have a build-rpms script?

File: tools/junitreport/README.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
+# junitreport
```
Comment (smarterclayton) [inline tools/junitreport/README.md]: You are no longer supposed to vendor this

File: pkg/boilerplate/operator/option.go [hunk 6]
```diff
@@ -0,0 +1,28 @@
+package operator
```
Comment (smarterclayton) [inline pkg/boilerplate/operator/option.go]: Why do you have a package called `boilerplate`?

File: Dockerfile [hunk 7]
```diff
@@ -1,8 +1,12 @@
-FROM registry.svc.ci.openshift.org/openshift/release:golang-1.10 AS builder
-COPY . /go/src/github.com/openshift/cluster-machine-appover
-RUN cd /go/src/github.com/openshift/cluster-machine-appover && go build -o machine-approver .
+#
```
Comment (smarterclayton) [inline Dockerfile]: This less consistent with our default pattern than the previous (which was also not correct)

See https://github.com/openshift/cluster-version-operator/blob/master/Dockerfile for the standard format, you should only have to change the RUN line in the builder and insert your custom steps after the COPY line in the runtime image.

**Comment by smarterclayton** [comment]
```
W1221 23:01:41.804742       1 cmd.go:127] Using insecure, self-signed certificates
I1221 23:01:41.805108       1 crypto.go:459] Generating new CA for cluster-machine-approver-signer@1545433301 cert, and key in /tmp/serving-cert-143867860/serving-signer.crt, /tmp/serving-cert-143867860/serving-signer.key
I1221 23:01:42.028512       1 crypto.go:536] Generating server certificate in /tmp/serving-cert-143867860/tls.crt, key in /tmp/serving-cert-143867860/tls.key
I1221 23:01:42.288075       1 observer_polling.go:37] Adding reactor for file "/var/run/configmaps/config/operator-config.yaml"
I1221 23:01:42.288156       1 observer_polling.go:37] Adding reactor for file "/var/run/secrets/serving-cert/tls.crt"
I1221 23:01:42.288179       1 observer_polling.go:37] Adding reactor for file "/var/run/secrets/serving-cert/tls.key"
I1221 23:01:42.289040       1 observer_polling.go:96] Starting file observer
panic: unable to obtain replicaset reference for events: Get https://127.0.0.1:6443/api/v1/namespaces/openshift-cluster-machine-approver/pods: dial tcp 127.0.0.1:6443: connect: connection refused

goroutine 1 [running]:
github.com/openshift/cluster-machine-approver/vendor/github.com/openshift/library-go/pkg/controller/controllercmd.(*ControllerBuilder).Run(0xc420532240, 0xc4200e20f0, 0xc4200941e0, 0x3, 0x3)
	/go/src/github.com/openshift/cluster-machine-approver/vendor/github.com/openshift/library-go/pkg/controller/controllercmd/builder.go:155 +0xabd
github.com/openshift/cluster-machine-approver/vendor/github.com/openshift/library-go/pkg/controller/controllercmd.(*ControllerCommandConfig).StartController(0xc4205f8840, 0xc420094120, 0x247c9d8, 0x247c9d8)
	/go/src/github.com/openshift/cluster-machine-approver/vendor/github.com/openshift/library-go/pkg/controller/controllercmd/cmd.go:172 +0x557
github.com/openshift/cluster-machine-approver/vendor/github.com/openshift/library-go/pkg/controller/controllercmd.(*ControllerCommandConfig).NewCommand.func1(0xc4200d7b80, 0xc4206966e0, 0x0, 0x2)
	/go/src/github.com/openshift/cluster-machine-approver/vendor/github.com/openshift/library-go/pkg/controller/controllercmd/cmd.go:67 +0x265
github.com/openshift/cluster-machine-approver/vendor/github.com/spf13/cobra.(*Command).execute(0xc4200d7b80, 0xc420696680, 0x2, 0x2, 0xc4200d7b80, 0xc420696680)
	/go/src/github.com/openshift/cluster-machine-approver/vendor/github.com/spf13/cobra/command.go:760 +0x2c1
github.com/openshift/cluster-machine-approver/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc4200d7900, 0x1, 0xc4200d7900, 0xc4200d7b80)
	/go/src/github.com/openshift/cluster-machine-approver/vendor/github.com/spf13/cobra/command.go:846 +0x30a
github.com/openshift/cluster-machine-approver/vendor/github.com/spf13/cobra.(*Command).Execute(0xc4200d7900, 0x181d990, 0x245b620)
	/go/src/github.com/openshift/cluster-machine-approver/vendor/github.com/spf13/cobra/command.go:794 +0x2b
main.main()
	/go/src/github.com/openshift/cluster-machine-approver/cmd/machine-approver/main.go:29 +0x187
```

in 
https://storage.googleapis.com/origin-ci-test/logs/release-openshift-origin-installer-e2e-aws-4.0/2708/artifacts/release-e2e-aws/pods/openshift-cluster-machine-approver_machine-approver-5bb5cb965d-28ljh_machine-approver-controller_previous.log.gz

After the restart it looks like you are renewing your lease incredibly often:

```
I1221 23:05:37.153030       1 leaderelection.go:209] successfully renewed lease openshift-cluster-machine-approver/cluster-machine-approver-lock
I1221 23:05:39.231941       1 leaderelection.go:209] successfully renewed lease openshift-cluster-machine-approver/cluster-machine-approver-lock
I1221 23:05:41.440786       1 leaderelection.go:209] successfully renewed lease openshift-cluster-machine-approver/cluster-machine-approver-lock
I1221 23:05:43.456531       1 leaderelection.go:209] successfully renewed lease openshift-cluster-machine-approver/cluster-machine-approver-lock
I1221 23:05:45.581277       1 leaderelection.go:209] successfully renewed lease openshift-cluster-machine-approver/cluster-machine-approver-lock
I1221 23:05:47.651659       1 leaderelection.go:209] successfully renewed lease openshift-cluster-machine-approver/cluster-machine-approver-lock
I1
```

Please fix that (and fix the library-go default if that's at fault).  This controller doesn't even need really need a lock.

File: pkg/boilerplate/operator/option.go (hunk 6 above)
Comment (enj) [reply pkg/boilerplate/operator/option.go]: It is my personal variation on the controller / operator sync loop.  It is used today by the following operators:

1. Alfred
1. Console
1. Authentication
1. Here

My goal was to add it to library-go once I felt it was "good enough."

File: tools/junitreport/README.md (hunk 5 above)
Comment (enj) [reply tools/junitreport/README.md]: Junk some SSCS.

File: hack/build-rpms.sh (hunk 4 above)
Comment (enj) [reply hack/build-rpms.sh]: Inert junk from SSCS.

File: Dockerfile (hunk 7 above)
Comment (enj) [reply Dockerfile]: :+1:

**Comment by enj** [comment]
> ```
> W1221 23:01:41.804742       1 cmd.go:127] Using insecure, self-signed certificates
> I1221 23:01:41.805108       1 crypto.go:459] Generating new CA for cluster-machine-approver-signer@1545433301 cert, and key in /tmp/serving-cert-143867860/serving-signer.crt, /tmp/serving-cert-143867860/serving-signer.key
> I1221 23:01:42.028512       1 crypto.go:536] Generating server certificate in /tmp/serving-cert-143867860/tls.crt, key in /tmp/serving-cert-143867860/tls.key
> I1221 23:01:42.288075       1 observer_polling.go:37] Adding reactor for file "/var/run/configmaps/config/operator-config.yaml"
> I1221 23:01:42.288156       1 observer_polling.go:37] Adding reactor for file "/var/run/secrets/serving-cert/tls.crt"
> I1221 23:01:42.288179       1 observer_polling.go:37] Adding reactor for file "/var/run/secrets/serving-cert/tls.key"
> I1221 23:01:42.289040       1 observer_polling.go:96] Starting file observer
> panic: unable to obtain replicaset reference for events: Get https://127.0.0.1:6443/api/v1/namespaces/openshift-cluster-machine-approver/pods: dial tcp 127.0.0.1:6443: connect: connection refused
> 
> goroutine 1 [running]:
> github.com/openshift/cluster-machine-approver/vendor/github.com/openshift/library-go/pkg/controller/controllercmd.(*ControllerBuilder).Run(0xc420532240, 0xc4200e20f0, 0xc4200941e0, 0x3, 0x3)
> 	/go/src/github.com/openshift/cluster-machine-approver/vendor/github.com/openshift/library-go/pkg/controller/controllercmd/builder.go:155 +0xabd
> github.com/openshift/cluster-machine-approver/vendor/github.com/openshift/library-go/pkg/controller/controllercmd.(*ControllerCommandConfig).StartController(0xc4205f8840, 0xc420094120, 0x247c9d8, 0x247c9d8)
> 	/go/src/github.com/openshift/cluster-machine-approver/vendor/github.com/openshift/library-go/pkg/controller/controllercmd/cmd.go:172 +0x557
> github.com/openshift/cluster-machine-approver/vendor/github.com/openshift/library-go/pkg/controller/controllercmd.(*ControllerCommandConfig).NewCommand.func1(0xc4200d7b80, 0xc4206966e0, 0x0, 0x2)
> 	/go/src/github.com/openshift/cluster-machine-approver/vendor/github.com/openshift/library-go/pkg/controller/controllercmd/cmd.go:67 +0x265
> github.com/openshift/cluster-machine-approver/vendor/github.com/spf13/cobra.(*Command).execute(0xc4200d7b80, 0xc420696680, 0x2, 0x2, 0xc4200d7b80, 0xc420696680)
> 	/go/src/github.com/openshift/cluster-machine-approver/vendor/github.com/spf13/cobra/command.go:760 +0x2c1
> github.com/openshift/cluster-machine-approver/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc4200d7900, 0x1, 0xc4200d7900, 0xc4200d7b80)
> 	/go/src/github.com/openshift/cluster-machine-approver/vendor/github.com/spf13/cobra/command.go:846 +0x30a
> github.com/openshift/cluster-machine-approver/vendor/github.com/spf13/cobra.(*Command).Execute(0xc4200d7900, 0x181d990, 0x245b620)
> 	/go/src/github.com/openshift/cluster-machine-approver/vendor/github.com/spf13/cobra/command.go:794 +0x2b
> main.main()
> 	/go/src/github.com/openshift/cluster-machine-approver/cmd/machine-approver/main.go:29 +0x187
> ```

I still do not understand the `127.0.0.1:6443` bit.  I would expect something like `ci-op-rx4tgcrl-3361d-api.origin-ci-int-aws.dev.rhcloud.com:6443`.

> in
> https://storage.googleapis.com/origin-ci-test/logs/release-openshift-origin-installer-e2e-aws-4.0/2708/artifacts/release-e2e-aws/pods/openshift-cluster-machine-approver_machine-approver-5bb5cb965d-28ljh_machine-approver-controller_previous.log.gz
> 
> After the restart it looks like you are renewing your lease incredibly often:
> 
> ```
> I1221 23:05:37.153030       1 leaderelection.go:209] successfully renewed lease openshift-cluster-machine-approver/cluster-machine-approver-lock
> I1221 23:05:39.231941       1 leaderelection.go:209] successfully renewed lease openshift-cluster-machine-approver/cluster-machine-approver-lock
> I1221 23:05:41.440786       1 leaderelection.go:209] successfully renewed lease openshift-cluster-machine-approver/cluster-machine-approver-lock
> I1221 23:05:43.456531       1 leaderelection.go:209] successfully renewed lease openshift-cluster-machine-approver/cluster-machine-approver-lock
> I1221 23:05:45.581277       1 leaderelection.go:209] successfully renewed lease openshift-cluster-machine-approver/cluster-machine-approver-lock
> I1221 23:05:47.651659       1 leaderelection.go:209] successfully renewed lease openshift-cluster-machine-approver/cluster-machine-approver-lock
> I1
> ```
> 
> Please fix that (and fix the library-go default if that's at fault). This controller doesn't even need really need a lock.

Yeah,

https://github.com/openshift/library-go/blob/077f461517a8617b5bdb5088afa7b451dd612b15/pkg/config/leaderelection/leaderelection.go#L83-L85

**Comment by enj** [comment]
Ah nvm, this runs with `hostNetwork: true` so the local IP makes sense.  I think we may just need a custom controller command to better account for this controller's requirements.

**Pull request merged**